### ISSUE TITLE
Mention of performance in comparison with Dhall?

### ISSUE THREAD

**Issue by arobertn** [tweag/nickel#560]
**Is your feature request related to a problem? Please describe.**

Nickel is pitched as a configuration language and compares itself to [Dhall](https://dhall-lang.org) in its [main README](https://github.com/tweag/nickel/blob/master/README.md) and [Rationale](https://github.com/tweag/nickel/blob/master/RATIONALE.md) documents. The comparison focuses on the need for type annotations and the strength of typing in Dhall, however Dhall also suffers from performance problems impacting scaling to larger configurations when working with, for example, Kubernetes:

• [Performance advice for dhall-kubernetes based package – dhall-haskell Issue #1890](https://github.com/dhall-lang/dhall-haskell/issues/1890)
• [Production usage – dhall-kubernetes Issue #157](https://github.com/dhall-lang/dhall-kubernetes/issues/157)
• [Slow (or even hanging) dhall-kubernetes based generation – dhall-haskell Issue #1960](https://github.com/dhall-lang/dhall-haskell/issues/1960)
• [Roadmap for improved Kubernetes support – Dhall discussion](https://discourse.dhall-lang.org/t/roadmap-for-improved-kubernetes-support/313)
• [Using env in import causing a huge performance hit – dhall-haskell Issue #1878](https://github.com/dhall-lang/dhall-haskell/issues/1878)
• ["dhall resolve" does too much – dhall-haskell Issue #2116](https://github.com/dhall-lang/dhall-haskell/issues/2116)
• [RFC: Integrity check modes · dhall-lang Issue #1185](https://github.com/dhall-lang/dhall-lang/issues/1185)

These problems accrue because the Dhall interpreter resolves every definition down to base primitives before executing. This causes exponential resource growth interpreting code where progressively functionality is built hierarchically on progressively complex / specialized utility functions, because definitions can be repeated at low levels and are then repeated multiplicatively at higher levels if functions that call them are themselves called multiple times. Similar expansion also occurs over _all branches_ of if-else statements, compounding the issue.

In a private project I work on, we will need to either abandon Dhall or rearchitect substantially due to the high memory and compile-time requirements the above brings. It would be interesting to know whether Nickel will improve this area or not.

**Describe the solution you'd like**

_If_ scalability is something that will be improved in Nickel, then the comparison to Dhall in the RATIONALE (and maybe the main README) should mention this point.

**Describe alternatives you've considered**

Remaining silent on the issue is another alternative. If Nickel does not, in fact, scale, projects may discover this the hard way after having committed to the technology.

**Additional context**

–

**Comment by yannham** [comment]
I think Nickel has the potential to scale well, and scalability is in some form part of the design goals, e.g. we've been taking parallelizability into account when thinking about language features, about incremental evaluation, and so on. What's more, there is no reason we would do like Dhall does, especially now in hindsight.

That being said, the efforts has been mainly on core language design and implementation until now. Once #426 lands (and maybe another PR on closurization), we should hopefully not do anything obviously overly naive and stupid anymore with respect to performances. But we can't say much effort have been put into optimizing the interpreter further right now. There's still a lot of room for improvement.

So, the answer is something like "yes, most probably, eventually" situation, but it may be a bit hard to advertise and substantiate right now.

**Comment by winitzki** [comment]
@arobertn  From the Dhall side I have some comments:

The "normal form explosion" is the consequence of two design choices: 1) Dhall will evaluate under lambda until the body of a lambda is in normal form, 2) Dhall will substitute all `let` expressions; a normal form in Dhall currently cannot have any `let` expressions anywhere.  So, `let` expressions will be substituted into the body of a lambda. `let x = whatever in λ(y : T) → if x then x else x` will paste `x` three times under lambda.

When developing a library of functions, a programmer will often start with small functions and build up to larger ones. When new functions reuse old functions a few times via `let`, the bodies of functions get copied multiple times over into new functions, which again get pasted multiple times over. The result is an exponential growth of the length of the normal form of functions.

This has been discussed a few times in the past in the Dhall community, but so far no consensus has been reached about how to resolve this problem.

In Dhall there is a workaround that mitigates the exponential growth (limiting but not eliminating it). For example, a normal form can become 500KB instead of 80MB after the workaround. The workaround is based on a corner case in evaluating a pattern-match. Let me know if this is still relevant and interesting for you to explore. https://github.com/dhall-lang/dhall-lang/issues/1385

**Comment by yannham** [comment]
Thanks for the insights @winitzki. This was more or less our understanding as well. In the current (very experimental) incremental evaluation, which implies to persist some evaluation cache across runs, we very explicitly plan to NOT go the Dhall route regarding caching. First, Nickel uses weak reduction, so we never reduce under lambdas. More importantly, my working idea right now is similar in spirit to snapshotting (an interesting subset of) the heap, which would preserve all of the sharing. Think of a moving collector collection, but going to a persistent storage instead of moving to another part of the heap.

**Comment by winitzki** [comment]
I wonder how Nickel handles normal forms and caching. In Dhall, the design is that a user can define a function in a module, and the normal form of that function is then "compiled" to a cached binary representation. The binary representation in Dhall is not any kind of compiled virtual machine code; it's just a more efficient representation of lambda-terms than textual. So, a huge normal form will still make everything slow (multi-megabyte normal forms in binary).

In Nickel, is a module compiled separately into some kind of standard representation?

**Comment by yannham** [comment]
> I wonder how Nickel handles normal forms and caching

So, the short answer currently is: it doesn't. There is no cross-evaluation caching happening at the time; what I mentioned is an ongoing experiment. The big difference I see is that in Nickel we'll basically follow the same representation as at runtime, where sharing is always represented as an indirection (code + environment, or thunk, or closure if you'd like). I think performing the substitution (inlining) is what's causing the explosion of the size of the normal form in Dhall. In Nickel it would basically be a subset of the thunk graph in the heap, and variables will never be actually substituted (just point indirectly to the value in an environment). So while the resulting cache might be large for large values (but, I feel like this is unavoidable?), it would be bounded by heap size at the end of the execution, which is likely smaller than when there are still computations happening.

How the code is represented is still up in the air, but I think it's basically whatever representation the virtual machine is handling at runtime: currently it's some form of AST, but in the future it might very well be bytecode.
